### DISCUSSION TITLE
Is there a panel / tab implementation for this theme?

### DISCUSSION THREAD

**Discussion by SeanKilleen** [mmistakes/minimal-mistakes#3230] [Q&A]
Asking because I could find it useful and I swore I saw it before, but I could definitely be wrong.

I'm going to be writing a tutorial where I have different snippets for different languages at each step along the way.

I'd love to show these in some sort of a tabbed panel style format, and was wondering if this theme happened to support such a thing. As I mentioned, I thought I saw this previously listed as a feature, but I could be misremembering.

If anyone has a definitive answer on this, let me know. Other recommendations on how to achieve my goal are also welcome. Thanks!

**Comment by iBug** [comment]
Consider [Material for MkDocs](https://squidfunk.github.io/mkdocs-material/). It has support for tabbed content and is often a better choice for (technical) documentations.

**Comment by SeanKilleen** [comment]
Thanks @iBug. In this case, I am doing a tutorial blog post where I showcase code. Normally when I do this, it is in only one language or technique, so hasn't been an issue in the past. This particular tutorial is showing how to do the same steps in multiple ways, which is why it would be useful. 

I agree that other tools are better for writing technical docs. I just happen to be doing a blog post. Perhaps I'll do separate posts and use includes for common snippets or something along those lines.

**Comment by qiujames** [comment]
Decided to implement tabbed code blocks as a custom _include.
See it here https://qiujames.github.io/tabbed-content-in-mm/#demo.
Feel free to poke in the repository to take a look at the code.
It's styled similarly to the masthead and is included similarly to how galleries work.

Could have issues since I haven't thoroughly tested it yet, but it seems to be pretty clean so far.

1. Add into your _includes/ directory the following [tablist](https://github.com/qiujames/qiujames.github.io/blob/main/_includes/tablist)
2. Add into your _sass/minimal-mistakes/ directory the following [_tablist.scss](https://github.com/qiujames/qiujames.github.io/blob/main/_sass/minimal-mistakes/_tablist.scss)
3. Update your _sass/minimal-mistakes.scss to contain the import
4. Add into assets/js/ the following [tablist-script.js](https://github.com/qiujames/qiujames.github.io/blob/main/assets/js/tablist-script.js)
5. Load the js file into your site by following the typical MM [workflow](https://mmistakes.github.io/minimal-mistakes/docs/javascript/#customizing) by adding it to _config.yml or in the the `<head>`

**Reply by djvill** [reply]
For future readers, I got this to work by doing the above plus:
1. Add into _includes/ [tablist_content](https://github.com/qiujames/qiujames.github.io/blob/main/_includes/tablist_content)
2. Adding the variables `$max-width`, `$nav-height`, `$nav-toggle-height`, & `$intro-transition` to [_variables.scss](https://github.com/qiujames/qiujames.github.io/blob/main/_sass/minimal-mistakes/_variables.scss)

I'm using [academicpages.github.io](https://github.com/academicpages/academicpages.github.io), a minimal-mistakes fork

**Comment by djvill** [comment]
Thanks for this great implementation, @qiujames! How would I implement the tab data if I want it to pull from .md files in a directory rather than post's YAML header? I'd prefer to author in .md because the content I want to write up is too long to comfortably squish into a YAML header. For example, `python.md` would look like:

````md
---
title: Python
---

```python
print("Hello World");
```

Additional area under the code content where one
can add text that is also formatted in **markdown**.
You can add [links](https://qiujames.github.io/) or even images!
![birdie!](/assets/images/favicon.ico)
````

**Reply by qiujames** [reply]
Amazing! I'm glad you were able to get it to work!

It's been over a year so I forget some of the details, but I think getting the content out from another markdown *should* be possible but possibly a little bit gross. Give me a bit while I explore using liquid includes to get it out from another md file.

**Reply by qiujames** [reply]
At a high level, I think you can use the includes tab https://jekyllrb.com/docs/includes/ here to try and insert the markdown. You may need to then pipe the output through a [markdownify filter](https://jekyllrb.com/docs/liquid/filters/#:~:text=bar%2C%20or%20baz-,Markdownify,-Convert%20a%20Markdown)

I think* that should suffice, but you'll need to massage the parameter/liquid argument passing and then also play around with the md file structure.

**Reply by qiujames** [reply]
Update: It's possible! You won't be able to have the liquid front matter (the section between the ---) to be rendered, but you can include that in the original page. I also think nested tablist's won't quite work with this setup

**Comment by qiujames** [comment]
@djvill Here's a working implementation of tabbed content where the markdown contents are derived from another file.

There is a slight limitation in that you can only write raw markdown in that supplemental doc, however, you can achieve the liquid front matter by just including the liquid front matter in the original document. 
https://github.com/qiujames/qiujames.github.io/commit/1fc9faeff3cb079618820e6ef4e2d5b6672efb86

I'll go and make another blog post that documents how this "content from a md" works and revert the changes to the original blog post, but I linked the commit that works by modifying the original blog post for now. You should see the change in the "Another Markdown" tab 

https://qiujames.github.io/tabbed-content-in-mm/#demo

At a high level, you'll want to do the following
1. Create a place to store your tabbed content md files, (I think a limitation with includes is that it has to be in the relative directory where the posts are made [0]), so it'll look something like `_posts/my_tabbed_content_dir`
2. Add your markdown contents under that dir, so something like `_posts/my_tabbed_content_dir/code.md`
3. In the posts YAML header, refer to the code.md in a tablist using

```yaml
- title: Another Markdown
  content_md: code.md
```
4. Modify tablist_content to have the 
`{% if tab.content_md %}
{% include_relative my_tabbed_content_dir/{{ tab.content_md }} %}
{% endif %}` section

Hopefully that works alright!

[0]: You can probably customize the include tag that jekyll has to read from a new directory if you really hate dirtying the _posts directory, but this implementation should be enough to get you started

**Reply by djvill** [reply]
Awesome! I wonder if the implementation could be spruced up even more by specifying a `tabdir` YAML parameter or something. What I'm thinking of is similar to how academicpages.github.io does (e.g.) the [publications page](https://github.com/academicpages/academicpages.github.io/blob/master/_pages/publications.md?plain=1), the meat of which is:

```liquid
{% for post in site.publications reversed %}
  {% include archive-single.html %}
{% endfor %}
```

What's also nice about that is that by using Liquid's `site.`, you get away from the no-`../` restriction for `include_relative`.

Anyway, I'll fiddle with this on [my fork of academicpages.github.io](https://github.com/djvill/academicpages.github.io/) (I was running into gem dependency issues with your repo on my machine, for some reason) and get back to you!

**Reply by djvill** [reply]
Just thinking about this some more, that'd probably be implemented via a slightly different version of `_includes/tablist` that takes a `dir` parameter and assumes metadata can be pulled from the YAML headers of the files in that directory. I suppose one could handle that via an `{% elsif ...` for `{% assign tablist ...` and the `{% for tab ...` loop, but on first blush, it strikes me as more trouble than it's worth (different data structures of "everything in YAML header" vs. "content in files and metadata in headers").

**Comment by djvill** [comment]
Good news @qiujames, I've got a working tablist implementation on my [fork](https://github.com/djvill/qiujames.github.io/tree/03ed6758b0bdc73f5ad40d94695178ba78976fa2) that uses [Jekyll collections](https://jekyllrb.com/docs/collections/)--in other words, a bunch of .md files in a directory. This is a more flexible implementation of tablists that allows for arbitrary content (i.e., not just a code block then a richtext block).

Within this framework, I still want to implement/figure out:
- nested tablists (might work off-the-shelf, might not)
- [custom sorting](https://jekyllrb.com/docs/collections/#custom-sorting-of-documents)
- if it's possible to get Liquid to work (maybe something to do with `content` vs. `output` [document-variables](https://jekyllrb.com/docs/collections/#documents))
- placing the collection folder in an arbitrary location (e.g., as a subfolder of `_posts/`, like in your "Another Markdown" example) 
- getting URIs to work for the tab content (`div.tablist-content`); currently this only works for the active tab

Thanks again for opening the door on this implementation!
